File: src/com/nerdscorner/android/plugin/github/ui/tablemodels/BaseModel.java
package com.nerdscorner.android.plugin.github.ui.tablemodels;

import java.io.Serializable;
import java.util.Comparator;
import java.util.List;

import javax.swing.table.AbstractTableModel;

import com.nerdscorner.android.plugin.github.domain.gh.Wrapper;
import com.nerdscorner.android.plugin.utils.ListUtils;

public abstract class BaseModel<T extends Wrapper> extends AbstractTableModel implements Serializable, Comparator<T> {
    public static final int COLUMN_NAME = 0;

    private final int colsCount;
    private final String[] colNames;

    final List<T> items;

    BaseModel(List<T> items, String[] colNames) {
        this.items = items;
        this.colNames = colNames;
        this.colsCount = colNames.length;
    }

    public void addRow(T item) {
        int row = items.size();
        items.add(item);
        items.sort(this);
        fireTableRowsInserted(row, row);
    }

    public void removeAllRows() {
        items.clear();
        fireTableDataChanged();
    }

    @Override
    public String getColumnName(int column) {
        return colNames[column];
    }

    @Override
    public int getRowCount() {
        return items.size();
    }

    @Override
    public int getColumnCount() {
        return colsCount;
    }

    @Override
    public boolean isCellEditable(int rowIndex, int columnIndex) {
        return false;
    }

    @Override
    public Class<?> getColumnClass(int columnIndex) {
        return String.class;
    }

    public T getRow(int rowIndex) {
        if (ListUtils.isEmpty(items)) {
            return null;
        }
        return items.get(rowIndex);
    }

    @Override
    public Object getValueAt(int rowIndex, int columnIndex) {
        if (rowIndex < 0 || rowIndex >= items.size()) {
            return null;
        }
        return items.get(rowIndex);
    }

    @Override
    public int compare(T one, T other) {
        return one.compare(other);
    }
}

File: src/com/nerdscorner/android/plugin/github/ui/tablemodels/BaseModel.kt
package com.nerdscorner.android.plugin.github.ui.tablemodels

import com.nerdscorner.android.plugin.github.domain.gh.Wrapper
import java.io.Serializable
import java.util.Comparator
import javax.swing.table.AbstractTableModel

abstract class BaseModel<T : Wrapper> internal constructor(internal val items: MutableList<T>, private val colNames: Array<String>)
    : AbstractTableModel(), Serializable, Comparator<T> {

    private val colsCount: Int = colNames.size

    fun addRow(item: T) {
        val row = items.size
        items.add(item)
        items.sortWith(this)
        fireTableRowsInserted(row, row)
    }

    fun removeAllRows() {
        items.clear()
        fireTableDataChanged()
    }

    override fun getColumnName(column: Int): String {
        return colNames[column]
    }

    override fun getRowCount(): Int {
        return items.size
    }

    override fun getColumnCount(): Int {
        return colsCount
    }

    override fun isCellEditable(rowIndex: Int, columnIndex: Int): Boolean {
        return false
    }

    override fun getColumnClass(columnIndex: Int): Class<*> {
        return String::class.java
    }

    fun getRow(rowIndex: Int): T? {
        return if (items.isEmpty()) {
            null
        } else items[rowIndex]
    }

    override fun getValueAt(rowIndex: Int, columnIndex: Int): Any? {
        return if (rowIndex < 0 || rowIndex >= items.size) {
            null
        } else items[rowIndex]
    }

    override fun compare(one: T, other: T): Int {
        return other.compare(one)
    }

    companion object {
        val COLUMN_NAME = 0
    }
}
